File: scripts/compat/windows-audit.py
#!/usr/bin/env python3
"""Read-only Windows compatibility audit for PIDEX.

This script reports local environment readiness signals only. It must not install,
start, delete, or modify anything.
"""

from __future__ import annotations

import argparse
import json
import os
import platform
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Any


COMMANDS = ("bash", "node", "npm", "python3", "python", "git", "pi")
RISKY_ENTRYPOINTS = (
    {
        "path": "install.sh",
        "status": "linux-owned",
        "windows_guidance": "Prefer a future install.windows.ps1; do not add inline Windows branches to install.sh.",
    },
    {
        "path": "uninstall.sh",
        "status": "linux-owned",
        "windows_guidance": "Prefer a future uninstall.windows.ps1.",
    },
    {
        "path": "dashboard/start.sh",
        "status": "linux-owned",
        "windows_guidance": "Prefer a future dashboard/start.windows.mjs or dashboard/start.windows.ps1.",
    },
    {
        "path": "scripts/release/public-readiness.sh",
        "status": "linux-owned",
        "windows_guidance": "Use this read-only audit or a future public-readiness.windows.py for Windows checks.",
    },
    {
        "path": "scripts/git-hooks/install-global.sh",
        "status": "linux-owned",
        "windows_guidance": "Use a future Windows-specific hook installer or document unsupported behavior.",
    },
)


def repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def git_bash_candidates() -> list[str]:
    candidates: list[str] = []
    for env_name in ("ProgramFiles", "ProgramFiles(x86)"):
        base = os.environ.get(env_name)
        if base:
            candidates.append(str(Path(base) / "Git" / "bin" / "bash.exe"))
    return candidates


def resolve_command(command: str) -> str | None:
    found = shutil.which(command)
    if found:
        return found
    if command == "bash" and platform.system().lower() == "windows":
        for candidate in git_bash_candidates():
            if Path(candidate).is_file():
                return candidate
    return None


def command_version(command: str, executable: str) -> tuple[str | None, bool]:
    version_args = {
        "bash": ["--version"],
        "node": ["--version"],
        "npm": ["--version"],
        "python3": ["--version"],
        "python": ["--version"],
        "git": ["--version"],
        "pi": ["--version"],
    }[command]
    try:
        completed = subprocess.run(
            [executable, *version_args],
            check=False,
            capture_output=True,
            text=True,
            timeout=5,
        )
    except (OSError, subprocess.SubprocessError):
        return None, False
    output = (completed.stdout or completed.stderr).strip().splitlines()
    version = output[0] if output else None
    if completed.returncode != 0:
        return version, False
    return version, True


def audit_commands() -> dict[str, dict[str, Any]]:
    result: dict[str, dict[str, Any]] = {}
    for command in COMMANDS:
        found = resolve_command(command)
        version: str | None = None
        executable_works = False
        if found:
            version, executable_works = command_version(command, found)
        result[command] = {
            "available": bool(found) and (executable_works or command == "pi"),
            "path": found,
            "version": version,
            "executable_works": executable_works,
        }
    return result


def detect_environment() -> dict[str, Any]:
    system = platform.system().lower()
    release = platform.release()
    env = {key: os.environ.get(key) for key in ("MSYSTEM", "MINGW_PREFIX", "WSL_DISTRO_NAME", "WT_SESSION") if os.environ.get(key)}
    is_wsl = bool(env.get("WSL_DISTRO_NAME")) or "microsoft" in release.lower()
    is_git_bash = bool(env.get("MSYSTEM") or env.get("MINGW_PREFIX"))

    if is_wsl:
        kind = "wsl"
    elif system == "windows" and is_git_bash:
        kind = "windows-git-bash"
    elif system == "windows":
        kind = "windows-native"
    elif is_git_bash:
        kind = "git-bash-like"
    elif system == "linux":
        kind = "linux"
    elif system == "darwin":
        kind = "macos"
    else:
        kind = system or "unknown"

    return {
        "kind": kind,
        "platform_system": platform.system(),
        "platform_release": release,
        "platform_version": platform.version(),
        "machine": platform.machine(),
        "python_executable": sys.executable,
        "environment_markers": env,
    }


def path_shape(root: Path) -> dict[str, Any]:
    root_text = str(root)
    home = str(Path.home())
    return {
        "repo_root": root_text,
        "home": home,
        "is_expected_linux_checkout": root_text == str(Path.home() / "pidex"),
        "contains_backslashes": "\\" in root_text,
        "contains_spaces": " " in root_text,
        "has_drive_prefix": len(root_text) >= 2 and root_text[1] == ":",
        "starts_with_slash": root_text.startswith("/"),
    }


def dashboard_prerequisites(root: Path, commands: dict[str, dict[str, Any]]) -> dict[str, Any]:
    dashboard = root / "dashboard"
    package_json = dashboard / "package.json"
    node_modules = dashboard / "node_modules"
    return {
        "dashboard_dir_exists": dashboard.is_dir(),
        "package_json_exists": package_json.is_file(),
        "node_available": commands["node"]["available"],
        "npm_available": commands["npm"]["available"],
        "node_modules_present": node_modules.is_dir(),
        "suggested_checks": [
            "npm --prefix dashboard run typecheck",
            "npm --prefix dashboard run build",
        ],
    }


def risky_entrypoints(root: Path) -> list[dict[str, Any]]:
    entries: list[dict[str, Any]] = []
    for item in RISKY_ENTRYPOINTS:
        path = root / item["path"]
        entries.append({**item, "exists": path.exists()})
    return entries


def parse_node_major_minor(version: str | None) -> tuple[int, int] | None:
    if not version:
        return None
    text = version.strip().lstrip("v")
    parts = text.split(".")
    try:
        return int(parts[0]), int(parts[1]) if len(parts) > 1 else 0
    except Exception:
        return None


def findings(environment: dict[str, Any], commands: dict[str, dict[str, Any]], paths: dict[str, Any]) -> list[dict[str, str]]:
    items: list[dict[str, str]] = []
    kind = environment["kind"]
    if kind == "linux":
        items.append({"level": "info", "message": "Linux detected; this is PIDEX's canonical currently tested path."})
    elif kind == "wsl":
        items.append({"level": "info", "message": "WSL detected; this is the safest Windows recommendation for now, pending explicit PIDEX smoke evidence."})
    elif kind == "windows-git-bash":
        items.append({"level": "warning", "message": "Windows Git Bash detected; Pi supports a Git Bash path, but PIDEX support is still experimental."})
    elif kind == "windows-native":
        items.append({"level": "warning", "message": "Native Windows detected; PIDEX does not currently claim PowerShell/CMD support."})
    else:
        items.append({"level": "warning", "message": f"Unvalidated platform detected: {kind}."})

    for command in ("bash", "node", "npm", "git", "pi"):
        if not commands[command]["available"]:
            items.append({"level": "warning", "message": f"Required or expected command not found on PATH: {command}."})
    if not (commands["python3"]["available"] or commands["python"]["available"]):
        items.append({"level": "warning", "message": "Neither python3 nor python was found on PATH."})

    node_version = parse_node_major_minor(commands["node"].get("version"))
    if node_version and node_version < (22, 12):
        items.append({"level": "warning", "message": "Dashboard dependencies require Node >=22.12.0; current node appears older."})

    if paths["contains_spaces"]:
        items.append({"level": "warning", "message": "PIDEX checkout path contains spaces; this is not yet validated for Windows support."})
    if not paths["is_expected_linux_checkout"]:
        items.append({"level": "info", "message": "PIDEX v0.1 documents ~/pidex as the expected checkout path."})

    return items


def build_report() -> dict[str, Any]:
    root = repo_root()
    environment = detect_environment()
    commands = audit_commands()
    paths = path_shape(root)
    return {
        "audit": "pidex-windows-compatibility",
        "read_only": True,
        "support_summary": {
            "linux": "supported/currently tested",
            "wsl2": "safest Windows recommendation for now",
            "windows_git_bash": "experimental/under analysis",
            "native_powershell_cmd": "not claimed",
        },
        "environment": environment,
        "commands": commands,
        "path_shape": paths,
        "dashboard": dashboard_prerequisites(root, commands),
        "known_unsupported_or_risky_entrypoints": risky_entrypoints(root),
        "findings": findings(environment, commands, paths),
    }


def print_text(report: dict[str, Any]) -> None:
    print("PIDEX Windows compatibility audit (read-only)")
    print("=" * 47)
    print(f"Environment: {report['environment']['kind']} ({report['environment']['platform_system']} {report['environment']['platform_release']})")
    print(f"Repo root: {report['path_shape']['repo_root']}")
    print("\nSupport summary:")
    for key, value in report["support_summary"].items():
        print(f"  - {key}: {value}")
    print("\nCommands:")
    for command, info in report["commands"].items():
        status = "found" if info["available"] else "missing"
        suffix = f" — {info['version']}" if info.get("version") else ""
        print(f"  - {command}: {status}{suffix}")
    print("\nDashboard prerequisites:")
    for key, value in report["dashboard"].items():
        if key != "suggested_checks":
            print(f"  - {key}: {value}")
    print("\nKnown unsupported/risky entrypoints:")
    for entry in report["known_unsupported_or_risky_entrypoints"]:
        print(f"  - {entry['path']}: {entry['status']} ({'exists' if entry['exists'] else 'missing'})")
    print("\nFindings:")
    for item in report["findings"]:
        print(f"  - [{item['level']}] {item['message']}")


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Read-only PIDEX Windows compatibility audit.")
    parser.add_argument("--json", action="store_true", help="print machine-readable JSON")
    args = parser.parse_args(argv)

    report = build_report()
    if args.json:
        print(json.dumps(report, indent=2, sort_keys=True))
    else:
        print_text(report)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/compat/windows-audit.mjs
#!/usr/bin/env node
// Read-only Windows compatibility audit for PIDEX. No installs, no mutation.

import { existsSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import { spawnSync } from 'node:child_process';

const COMMANDS = ['bash', 'node', 'npm', 'git', 'pi'];
const RISKY_ENTRYPOINTS = [
  ['install.sh', 'linux-owned', 'Prefer a future install.windows.ps1; do not add inline Windows branches to install.sh.'],
  ['uninstall.sh', 'linux-owned', 'Prefer a future uninstall.windows.ps1.'],
  ['dashboard/start.sh', 'linux-owned', 'Prefer a future dashboard/start.windows.mjs or dashboard/start.windows.ps1.'],
  ['scripts/release/public-readiness.sh', 'linux-owned', 'Use this read-only audit or a future public-readiness.windows.mjs for Windows checks.'],
  ['scripts/git-hooks/install-global.sh', 'linux-owned', 'Use a future Windows-specific hook installer or document unsupported behavior.'],
];

function repoRoot() {
  return path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');
}

function splitPathEnv() {
  return (process.env.Path || process.env.PATH || '').split(path.delimiter).filter(Boolean);
}

function commandNames(command) {
  if (process.platform === 'win32') {
    if (command.endsWith('.exe') || command.endsWith('.cmd') || command.endsWith('.ps1')) return [command];
    return [`${command}.exe`, `${command}.cmd`, `${command}.ps1`, command];
  }
  return [command];
}

function which(command) {
  for (const dir of splitPathEnv()) {
    for (const name of commandNames(command)) {
      const candidate = path.join(dir, name);
      if (existsSync(candidate)) return candidate;
    }
  }
  return null;
}

function gitBashCandidates() {
  return ['ProgramFiles', 'ProgramFiles(x86)']
    .map((key) => process.env[key])
    .filter(Boolean)
    .map((base) => path.join(base, 'Git', 'bin', 'bash.exe'));
}

function resolveCommand(command) {
  const found = which(command);
  if (found) return found;
  if (command === 'bash' && process.platform === 'win32') {
    return gitBashCandidates().find((candidate) => existsSync(candidate)) || null;
  }
  return null;
}

function commandVersion(command, executable) {
  const argsByCommand = {
    bash: ['--version'],
    node: ['--version'],
    npm: ['--version'],
    git: ['--version'],
    pi: ['--version'],
  };
  let runCommand = executable;
  let runArgs = argsByCommand[command];
  if (process.platform === 'win32') {
    const lower = executable.toLowerCase();
    if (['npm', 'pi'].includes(command)) {
      runCommand = process.env.ComSpec || 'cmd.exe';
      runArgs = ['/d', '/c', `${command} ${argsByCommand[command].join(' ')}`];
    } else if (lower.endsWith('.ps1')) {
      runCommand = process.env.SystemRoot ? path.join(process.env.SystemRoot, 'System32', 'WindowsPowerShell', 'v1.0', 'powershell.exe') : 'powershell.exe';
      runArgs = ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', executable, ...argsByCommand[command]];
    } else if (lower.endsWith('.cmd') || lower.endsWith('.bat') || !path.extname(executable)) {
      runCommand = process.env.ComSpec || 'cmd.exe';
      runArgs = ['/d', '/c', `${command} ${argsByCommand[command].join(' ')}`];
    }
  }
  const result = spawnSync(runCommand, runArgs, { encoding: 'utf8', timeout: 5000 });
  const output = `${result.stdout || ''}${result.stderr || ''}`.trim().split(/\r?\n/).filter(Boolean);
  return { version: output[0] || null, works: result.status === 0 };
}

function auditCommands() {
  const result = {};
  for (const command of COMMANDS) {
    const found = resolveCommand(command);
    let version = null;
    let executableWorks = false;
    if (found) {
      ({ version, works: executableWorks } = commandVersion(command, found));
    }
    result[command] = {
      available: Boolean(found) && (executableWorks || command === 'pi'),
      path: found,
      version,
      executable_works: executableWorks,
    };
  }
  return result;
}

function detectEnvironment() {
  const release = os.release();
  const env = Object.fromEntries(['MSYSTEM', 'MINGW_PREFIX', 'WSL_DISTRO_NAME', 'WT_SESSION'].filter((k) => process.env[k]).map((k) => [k, process.env[k]]));
  const isWsl = Boolean(env.WSL_DISTRO_NAME) || release.toLowerCase().includes('microsoft');
  const isGitBash = Boolean(env.MSYSTEM || env.MINGW_PREFIX);
  let kind;
  if (isWsl) kind = 'wsl';
  else if (process.platform === 'win32' && isGitBash) kind = 'windows-git-bash';
  else if (process.platform === 'win32') kind = 'windows-native';
  else if (isGitBash) kind = 'git-bash-like';
  else if (process.platform === 'linux') kind = 'linux';
  else if (process.platform === 'darwin') kind = 'macos';
  else kind = process.platform || 'unknown';
  return {
    kind,
    platform_system: os.type(),
    platform_release: release,
    platform_version: os.version?.() || '',
    machine: os.arch(),
    node_executable: process.execPath,
    environment_markers: env,
  };
}

function pathShape(root) {
  const home = os.homedir();
  return {
    repo_root: root,
    home,
    is_expected_linux_checkout: path.resolve(root).toLowerCase() === path.resolve(home, 'pidex').toLowerCase(),
    contains_backslashes: root.includes('\\'),
    contains_spaces: root.includes(' '),
    has_drive_prefix: /^[A-Za-z]:/.test(root),
    starts_with_slash: root.startsWith('/'),
  };
}

function dashboardPrerequisites(root, commands) {
  const dashboard = path.join(root, 'dashboard');
  return {
    dashboard_dir_exists: existsSync(dashboard),
    package_json_exists: existsSync(path.join(dashboard, 'package.json')),
    node_available: commands.node.available,
    npm_available: commands.npm.available,
    node_modules_present: existsSync(path.join(dashboard, 'node_modules')),
    suggested_checks: ['npm --prefix dashboard run typecheck', 'npm --prefix dashboard run build'],
  };
}

function riskyEntrypoints(root) {
  return RISKY_ENTRYPOINTS.map(([entryPath, status, windows_guidance]) => ({
    path: entryPath,
    status,
    windows_guidance,
    exists: existsSync(path.join(root, ...entryPath.split('/'))),
  }));
}

function parseNodeMajorMinor(version) {
  if (!version) return null;
  const match = version.trim().match(/^v?(\d+)\.(\d+)/);
  return match ? [Number(match[1]), Number(match[2])] : null;
}

function findings(environment, commands, paths) {
  const items = [];
  if (environment.kind === 'linux') items.push({ level: 'info', message: "Linux detected; this is PIDEX's canonical currently tested path." });
  else if (environment.kind === 'wsl') items.push({ level: 'info', message: 'WSL detected; this is the safest Windows recommendation for now, pending explicit PIDEX smoke evidence.' });
  else if (environment.kind === 'windows-git-bash') items.push({ level: 'warning', message: 'Windows Git Bash detected; Pi supports a Git Bash path, but PIDEX support is still experimental.' });
  else if (environment.kind === 'windows-native') items.push({ level: 'warning', message: 'Native Windows detected; PIDEX does not currently claim full PowerShell/CMD runtime support.' });
  else items.push({ level: 'warning', message: `Unvalidated platform detected: ${environment.kind}.` });

  for (const command of ['bash', 'node', 'npm', 'git', 'pi']) {
    if (!commands[command].available) items.push({ level: 'warning', message: `Required or expected command not found on PATH/standard location: ${command}.` });
  }
  const nodeVersion = parseNodeMajorMinor(commands.node.version);
  if (nodeVersion && (nodeVersion[0] < 22 || (nodeVersion[0] === 22 && nodeVersion[1] < 12))) {
    items.push({ level: 'warning', message: 'PIDEX Windows bootstrap/dashboard dependencies require Node >=22.12.0; current node appears older.' });
  }
  if (paths.contains_spaces) items.push({ level: 'warning', message: 'PIDEX checkout path contains spaces; this is not yet validated for Windows support.' });
  if (!paths.is_expected_linux_checkout) items.push({ level: 'info', message: 'PIDEX v0.1 documents ~/pidex / $HOME\\pidex as the expected checkout path.' });
  return items;
}

function buildReport() {
  const root = repoRoot();
  const environment = detectEnvironment();
  const commands = auditCommands();
  const paths = pathShape(root);
  return {
    audit: 'pidex-windows-compatibility',
    implementation: 'node',
    read_only: true,
    support_summary: {
      linux: 'supported/currently tested',
      wsl2: 'safest Windows recommendation for now',
      windows_git_bash: 'experimental/under analysis',
      native_powershell_cmd: 'experimental bootstrap only; full runtime not claimed',
    },
    environment,
    commands,
    path_shape: paths,
    dashboard: dashboardPrerequisites(root, commands),
    known_unsupported_or_risky_entrypoints: riskyEntrypoints(root),
    findings: findings(environment, commands, paths),
  };
}

function printText(report) {
  console.log('PIDEX Windows compatibility audit (read-only)');
  console.log('===============================================');
  console.log(`Environment: ${report.environment.kind} (${report.environment.platform_system} ${report.environment.platform_release})`);
  console.log(`Repo root: ${report.path_shape.repo_root}`);
  console.log('\nSupport summary:');
  for (const [key, value] of Object.entries(report.support_summary)) console.log(`  - ${key}: ${value}`);
  console.log('\nCommands:');
  for (const [command, info] of Object.entries(report.commands)) {
    const status = info.available ? 'found' : 'missing';
    const suffix = info.version ? ` — ${info.version}` : '';
    console.log(`  - ${command}: ${status}${suffix}`);
  }
  console.log('\nDashboard prerequisites:');
  for (const [key, value] of Object.entries(report.dashboard)) if (key !== 'suggested_checks') console.log(`  - ${key}: ${value}`);
  console.log('\nKnown unsupported/risky entrypoints:');
  for (const entry of report.known_unsupported_or_risky_entrypoints) console.log(`  - ${entry.path}: ${entry.status} (${entry.exists ? 'exists' : 'missing'})`);
  console.log('\nFindings:');
  for (const item of report.findings) console.log(`  - [${item.level}] ${item.message}`);
}

const json = process.argv.includes('--json');
const report = buildReport();
if (json) console.log(JSON.stringify(report, null, 2));
else printText(report);
